### ISSUE TITLE
Missing documentation for register-custom-form-control directive

### ISSUE THREAD

**Issue by jonsamwell** [jonsamwell/angular-auto-validate#34]
Missing documentation for register-custom-form-control directive

**Comment by pjwalmsley** [comment]
I should/may be able to get around to this this weekend.  

I also struggled a bit with getting a nested form to validate.  It may be worth noting in the documentation that the inner form must have an element type of ng-form.  While this isn't an issue with your code it isn't blatantly obvious. :)

**Comment by jonsamwell** [comment]
Thanks, I'll add the bit about a nested ng-form to the docs!

**Comment by pjwalmsley** [comment]
Thanks for all your work on this library.  Another small typo i spotted: "you can still do that by just specifying which element modifier should be used to modifier the elements visual state" - 'modifier' should be 'modify'. :)

**Comment by jonsamwell** [comment]
Thanks - I'll update soon

**Comment by jonsamwell** [comment]
Typo updated

**Comment by pjwalmsley** [comment]
Worth noting or changing the limit of ten elements deep.  I've got a really deep form and this threw me off.

The recursion limit of ten elements is on line 1023 if you feel it is worth increasing or making recursve based on the existence of a parent.

**Comment by jonsamwell** [comment]
Maybe a config options i.e. recursionDepth

**Comment by pjwalmsley** [comment]
That would work.  I mean there really isn't much danger/speed loss when its higher.  

Of better yet if we used a recursive function there would be no need to worry about recursion depth.  Perhaps we could replace this function with the following.

            var findParentForm = function (el) {
                var form;
                var recursive = function (parent) {
                    if (parent !== undefined) {
                        if (parent.nodeName.toLowerCase() === 'form') {
                            form = parent;
                        } else {
                            recursive(angular.element(parent).parent()[0]);
                        }
                    }
                }
                recursive(el);

**Comment by pjwalmsley** [comment]
Just sent over a pull request.

**Comment by pjwalmsley** [comment]
So i messed up that pull request.  The change, as described above, can be seen on my branch.  It's commit  0908f71 .

**Comment by hinok** [comment]
Hey,
Can someone provide very basic example why we need this directive?
I use <my-custom-directive> inside a form and everything is still propery validated.

plnkr http://plnkr.co/edit/GMoTi673v8cCR2UwQPRf?p=preview

**Comment by pjwalmsley** [comment]
Standard elements such as input, select, textfield, or button will be tracked by angular-auto-validate.  Custom fields that may provide a better user experience are not picked up (we find by tag type rather than ng-model).  

In my case I have a custom multi-select drop down, a field that combines/controls two input fields to exposes a single usable value, and a special field that can be cleared.

It is not needed in your example. :)

**Comment by hinok** [comment]
To be honest, I still don't get it.
For example I have my own file-upload directive which inject some upload validation errors into form.

In template it looks like this

```
<form name="userForm">
    <file-upload name="userAvatar" validation-max-size="1024"></file-upload>
    <button>Submit</button>
</form>
```

This directive will do something like this in case of uploading too big image
```
userForm.userAvatar.$setValidity('validationMaxSize', false)
```

As I understood, I could use `register-custom-form-control` to display errors?
How should I do it? Should I add `input hidden` with proper name inside file-upload's template?

**Comment by pjwalmsley** [comment]
Nope, your on the right track.  Auto validate's role is to add the error message, not necessarily do the validation.  

If you sought to validate that you would add the <register-custom-form-control>.  This watches for error conditions (such as validationMaxSize).  Then you would add a value for validationMaxSize to your lang file (see examples for errors like required or minLength).  

To actually have the error condition thrown you probably need to write a validation directive.  It looks like you may have already done this.  For a more traditional validation it will look something like:

    myApp.directive("validateMaxSizeOfModel", function(){
          // requires an isloated model
          return {
           // restrict to an attribute type.
           restrict: 'A',
          // element must have ng-model attribute.
           require: 'ngModel',
           link: function(scope, ele, attrs, ctrl){
              // add a parser that will process each time the value is
              // parsed into the model when the user updates it.
              ctrl.$parsers.unshift(function(value) {
                if(value == attrs.validateMaxSizeOfModel){
                  ctrl.$setValidity('validationMaxSize', valid);
                } else {

}

                // if it's valid, return the value to the model,
                // otherwise return undefined.
                return valid ? value : undefined;
              });

           }
          }
        });

**Comment by hinok** [comment]
@pjwalmsley Thank you! I changed *elementModifier a bit and problably this is the reason why I couldn't see any appended validation errors in <file-upload/> element while using `register-custom-form-control`. All is clear now, thank you again.
